### PR TITLE
refine some interface for better expose api.

### PR DESCRIPTION
**Pull request by guolinke**

### CODE REVIEW

File: include/LightGBM/config.h [hunk 1]
```diff
@@ -121,9 +121,9 @@ struct IOConfig: public ConfigBase {
 struct ObjectiveConfig: public ConfigBase {
 public:
   virtual ~ObjectiveConfig() {}
-  double sigmoid = 1;
+  float sigmoid = 1;
```
Comment (xuehui1991) [inline include/LightGBM/config.h]: float sigmoid = 1.0f ?

File: include/LightGBM/config.h [hunk 2]
```diff
@@ -137,9 +137,9 @@ struct MetricConfig: public ConfigBase {
   virtual ~MetricConfig() {}
   int early_stopping_round = 0;
   int output_freq = 1;
-  double sigmoid = 1;
+  float sigmoid = 1;
```
Comment (xuehui1991) [inline include/LightGBM/config.h]: float sigmoid = 1.0f ?

File: include/LightGBM/config.h [hunk 3]
```diff
@@ -149,13 +149,13 @@ struct MetricConfig: public ConfigBase {
 struct TreeConfig: public ConfigBase {
 public:
   int min_data_in_leaf = 100;
-  double min_sum_hessian_in_leaf = 10.0f;
+  float min_sum_hessian_in_leaf = 10.0f;
   // should > 1, only one leaf means not need to learning
   int num_leaves = 127;
   int feature_fraction_seed = 2;
-  double feature_fraction = 1.0;
+  float feature_fraction = 1.0;
```
Comment (xuehui1991) [inline include/LightGBM/config.h]: float feature_fraction = 1.0f ?

File: include/LightGBM/config.h [hunk 4]
```diff
@@ -149,13 +149,13 @@ struct MetricConfig: public ConfigBase {
...
   // max cache size(unit:MB) for historical histogram. < 0 means not limit
-  double histogram_pool_size = -1;
+  float histogram_pool_size = -1;
```
Comment (xuehui1991) [inline include/LightGBM/config.h]: float histogram_pool_size = -1.0f?

File: include/LightGBM/boosting.h [hunk 5]
```diff
@@ -28,12 +28,12 @@ class Boosting {
...
-  * \param output_model_filename Filename of output model
   */
-  virtual void Init(const Dataset* train_data,
+  virtual void Init(
+    const BoostingConfig* config,
+    const Dataset* train_data,
     const ObjectiveFunction* object_function,
-    const std::vector<const Metric*>& training_metrics,
-    const char* output_model_filename) = 0;
+    const std::vector<const Metric*>& training_metrics) = 0;
```
Comment (xuehui1991) [inline include/LightGBM/boosting.h]: do you change the training_metrics in Init function? I not sure using "&" here is perfect...

File: include/LightGBM/utils/common.h [hunk 6]
```diff
@@ -250,36 +248,23 @@ inline static void StringToIntArray(const std::string& str, char delimiter, size
...
-}
-
-inline static std::vector<double> StringToDoubleArray(const std::string& str, char delimiter) {
-  std::vector<std::string> strs = Split(str.c_str(), delimiter);
-  std::vector<double> ret;
+  std::vector<float> ret;
   for (size_t i = 0; i < strs.size(); ++i) {
     strs[i] = Trim(strs[i]);
-    double val = 0.0;
+    float val = 0.0;
```
Comment (xuehui1991) [inline include/LightGBM/utils/common.h]: float val = 0.0f ?

File: src/application/application.cpp [hunk 7]
```diff
@@ -240,37 +237,41 @@ void Application::InitTrain() {
...
+  boosting_->SaveModelToFile(is_finished, config_.io_config.output_model.c_str());
+  Log::Info("Finished train");
 }
 
 
 void Application::Predict() {
   // create predictor
-  Predictor predictor(boosting_, config_.io_config.is_sigmoid, config_.predict_leaf_index);
+  Predictor predictor(boosting_, config_.io_config.is_sigmoid, 
+    config_.predict_leaf_index, config_.io_config.num_model_predict);
```
Comment (xuehui1991) [inline src/application/application.cpp]: why a blank line here?

File: src/boosting/boosting.cpp [hunk 8]
```diff
@@ -3,13 +3,48 @@
...
+Boosting* Boosting::CreateBoosting(BoostingType type, const char* filename) {
+  if (filename[0] == '\0') {
+    if (type == BoostingType::kGBDT) {
+      return new GBDT();
+    } else {
+      return nullptr;
+    }
   } else {
-    return nullptr;
+    Boosting* ret = nullptr;
```
Comment (xuehui1991) [inline src/boosting/boosting.cpp]: This part looks like CreateBoosting(const char* filename), can we reuse that part?

File: src/io/config.cpp [hunk 9]
```diff
@@ -224,14 +244,14 @@ void MetricConfig::Set(const std::unordered_map<std::string, std::string>& param
...
   GetInt(params, "min_data_in_leaf", &min_data_in_leaf);
-  GetDouble(params, "min_sum_hessian_in_leaf", &min_sum_hessian_in_leaf);
+  GetFloat(params, "min_sum_hessian_in_leaf", &min_sum_hessian_in_leaf);
   CHECK(min_sum_hessian_in_leaf > 1.0f || min_data_in_leaf > 0);
   GetInt(params, "num_leaves", &num_leaves);
   CHECK(num_leaves > 1);
   GetInt(params, "feature_fraction_seed", &feature_fraction_seed);
-  GetDouble(params, "feature_fraction", &feature_fraction);
+  GetFloat(params, "feature_fraction", &feature_fraction);
   CHECK(feature_fraction > 0.0 && feature_fraction <= 1.0);
```
Comment (xuehui1991) [inline src/io/config.cpp]: feature_fraction <= 1.0f?

File: src/io/config.cpp [hunk 10]
```diff
@@ -243,9 +263,9 @@ void BoostingConfig::Set(const std::unordered_map<std::string, std::string>& par
   GetInt(params, "bagging_seed", &bagging_seed);
   GetInt(params, "bagging_freq", &bagging_freq);
   CHECK(bagging_freq >= 0);
-  GetDouble(params, "bagging_fraction", &bagging_fraction);
+  GetFloat(params, "bagging_fraction", &bagging_fraction);
   CHECK(bagging_fraction > 0.0 && bagging_fraction <= 1.0);
```
Comment (xuehui1991) [inline src/io/config.cpp]: bagging_fraction <= 1.0f

File: src/io/dataset.cpp [hunk 11]
```diff
@@ -593,8 +593,8 @@ void Dataset::ExtractFeaturesFromFile() {
   std::function<void(data_size_t, const std::vector<std::string>&)> process_fun =
     [this, &init_score]
   (data_size_t start_idx, const std::vector<std::string>& lines) {
-    std::vector<std::pair<int, double>> oneline_features;
-    double tmp_label = 0.0;
+    std::vector<std::pair<int, float>> oneline_features;
+    float tmp_label = 0.0;
     #pragma omp parallel for schedule(static) private(oneline_features) firstprivate(tmp_label)
     for (data_size_t i = 0; i < static_cast<data_size_t>(lines.size()); i++) {
```
Comment (xuehui1991) [inline src/io/dataset.cpp]: ++i？

File: src/io/dataset.cpp [hunk 12]
```diff
@@ -507,8 +507,8 @@ void Dataset::LoadValidationData(const Dataset* train_set, bool use_two_round_lo
 }
 
 void Dataset::ExtractFeaturesFromMemory() {
-  std::vector<std::pair<int, double>> oneline_features;
-  double tmp_label = 0.0;
+  std::vector<std::pair<int, float>> oneline_features;
+  float tmp_label = 0.0;
```
Comment (xuehui1991) [inline src/io/dataset.cpp]: float tmp_label = 0.0f ?

File: src/io/dataset.cpp [hunk 13]
```diff
@@ -593,8 +593,8 @@ void Dataset::ExtractFeaturesFromFile() {
   std::function<void(data_size_t, const std::vector<std::string>&)> process_fun =
     [this, &init_score]
   (data_size_t start_idx, const std::vector<std::string>& lines) {
-    std::vector<std::pair<int, double>> oneline_features;
-    double tmp_label = 0.0;
+    std::vector<std::pair<int, float>> oneline_features;
+    float tmp_label = 0.0;
```
Comment (xuehui1991) [inline src/io/dataset.cpp]: float tmp_label = 0.0f ?

Comment (xuehui1991) [inline src/io/dataset.cpp]: float tmp_label = 0.0f ?

File: src/io/config.cpp [hunk 14]
```diff
@@ -224,14 +244,14 @@ void MetricConfig::Set(const std::unordered_map<std::string, std::string>& param
...
   CHECK(min_sum_hessian_in_leaf > 1.0f || min_data_in_leaf > 0);
   GetInt(params, "num_leaves", &num_leaves);
   CHECK(num_leaves > 1);
   GetInt(params, "feature_fraction_seed", &feature_fraction_seed);
-  GetDouble(params, "feature_fraction", &feature_fraction);
+  GetFloat(params, "feature_fraction", &feature_fraction);
   CHECK(feature_fraction > 0.0 && feature_fraction <= 1.0);
-  GetDouble(params, "histogram_pool_size", &histogram_pool_size);
+  GetFloat(params, "histogram_pool_size", &histogram_pool_size);
   GetInt(params, "max_depth", &max_depth);
```
Comment (xuehui1991) [inline src/io/config.cpp]: please change this name "histogram_pool_size" if it is a float parameter...

File: src/boosting/gbdt.cpp [hunk 15]
```diff
@@ -392,38 +427,47 @@ void GBDT::FeatureImportance(const int last_iter) {
...
 }
 
-double GBDT::PredictRaw(const double* value) const {
-  double ret = 0.0;
-  for (size_t i = 0; i < models_.size(); ++i) {
+float GBDT::PredictRaw(const float* value, int num_used_model) const {
+  if (num_used_model < 0) {
+    num_used_model = static_cast<int>(models_.size());
+  }
+  float ret = 0.0;
```
Comment (xuehui1991) [inline src/boosting/gbdt.cpp]: float ret = 0.0f

File: include/LightGBM/utils/common.h [hunk 16]
```diff
@@ -237,6 +237,20 @@ inline static std::string ArrayToString(const T* arr, int n, char delimiter) {
   return ss.str();
 }
 
+template<typename T>
+inline static std::string ArrayToString(std::vector<T> arr, char delimiter) {
+  if (arr.size() <= 0) {
+    return std::string("");
+  }
+  std::stringstream ss;
```
Comment (xuehui1991) [inline include/LightGBM/utils/common.h]: change the name?

File: src/io/config.cpp [hunk 17]
```diff
@@ -269,6 +265,9 @@ void BoostingConfig::Set(const std::unordered_map<std::string, std::string>& par
   CHECK(learning_rate > 0.0);
```
Comment (xuehui1991) [inline src/io/config.cpp]: CHECK(learning_rate > 0.0f) ?

File: src/metric/rank_metric.hpp [hunk 18]
```diff
@@ -60,13 +62,12 @@ class NDCGMetric:public Metric {
...
-                               query_boundaries_[i + 1] - query_boundaries_[i],
-                               &inverse_max_dcgs_[i]);
+        query_boundaries_[i + 1] - query_boundaries_[i],
+        &inverse_max_dcgs_[i]);
       for (size_t j = 0; j < inverse_max_dcgs_[i].size(); ++j) {
         if (inverse_max_dcgs_[i][j] > 0.0f) {
           inverse_max_dcgs_[i][j] = 1.0f / inverse_max_dcgs_[i][j];
-        }
-        else {
+        } else {
```
Comment (xuehui1991) [inline src/metric/rank_metric.hpp]: why change this?

**Comment by chivee** [comment]
refer to #34
